File: skellig-test-step-processing/src/main/java/org/skellig/teststep/processing/validation/DefaultTestStepResultValidator.java
package org.skellig.teststep.processing.validation;

import org.skellig.teststep.processing.exception.ValidationException;
import org.skellig.teststep.processing.model.ExpectedResult;
import org.skellig.teststep.processing.model.ValidationType;
import org.skellig.teststep.processing.validation.comparator.ValueComparator;
import org.skellig.teststep.processing.valueextractor.TestStepValueExtractor;

import java.util.Arrays;
import java.util.Collection;
import java.util.List;
import java.util.stream.Collectors;

public class DefaultTestStepResultValidator implements TestStepResultValidator {

    private ValueComparator valueComparator;
    private TestStepValueExtractor valueExtractor;

    protected DefaultTestStepResultValidator(ValueComparator valueComparator,
                                             TestStepValueExtractor valueExtractor) {
        this.valueComparator = valueComparator;
        this.valueExtractor = valueExtractor;
    }

    public void validate(ExpectedResult expectedResult, Object actualResult) {
        StringBuilder errorBuilder = new StringBuilder();
        if (!validate(expectedResult, actualResult, errorBuilder)) {
            throw new ValidationException("Validation failed!\n" + errorBuilder.toString());
        }
    }

    private boolean validate(ExpectedResult expectedResult, Object actualResult, StringBuilder errorBuilder) {
        if (expectedResult.getValidationType() == ValidationType.ANY_MATCH) {
            return expectedResult.<List<ExpectedResult>>getExpectedResult().stream()
                    .anyMatch(item -> validate2(item, actualResult, errorBuilder));
        } else if (expectedResult.getValidationType() == ValidationType.ANY_NONE_MATCH) {
            return expectedResult.<List<ExpectedResult>>getExpectedResult().stream()
                    .anyMatch(item -> !validate2(item, actualResult, errorBuilder));
        } else if (expectedResult.getValidationType() == ValidationType.NONE_MATCH) {
            return expectedResult.<List<ExpectedResult>>getExpectedResult().stream()
                    .noneMatch(item -> validate2(item, actualResult, errorBuilder));
        } else if (expectedResult.getValidationType() == ValidationType.ALL_MATCH) {
            return expectedResult.<List<ExpectedResult>>getExpectedResult().stream()
                    .allMatch(item -> validate2(item, actualResult, errorBuilder));
        } else {
            boolean isValid = valueComparator.compare(expectedResult.getExpectedResult(), actualResult);

            if (expectedResult.getValidationTypeOfParent() == ValidationType.NONE_MATCH && isValid ||
                    expectedResult.getValidationTypeOfParent() != ValidationType.NONE_MATCH && !isValid) {
                constructErrorMessage(expectedResult, actualResult, errorBuilder);
            }
            return isValid;
        }
    }

    private boolean validate2(ExpectedResult expectedResult, Object actualResult, StringBuilder errorBuilder) {
        Object actualValue = extractActualValueFromExpectedResult(actualResult, expectedResult);
        if (expectedResult.getProperty() == null && actualValue instanceof Collection) {
            if (expectedResult.getValidationType() == ValidationType.NONE_MATCH) {
                return ((Collection) actualValue).stream().allMatch(actualResultItem ->
                        validate(expectedResult, actualResultItem, errorBuilder));
            } else {
                return ((Collection) actualValue).stream().anyMatch(actualResultItem ->
                        validate(expectedResult, actualResultItem, errorBuilder));
            }
        } else if (actualResult.getClass().isArray()) {
            return validate2(expectedResult, Arrays.stream((Object[]) actualResult).collect(Collectors.toList()), errorBuilder);
        } else {
            return validate(expectedResult, actualValue, errorBuilder);
        }
    }

    private Object extractActualValueFromExpectedResult(Object actualResult, Object expectedResult) {
        if (ExpectedResult.class.equals(expectedResult.getClass())) {
            return valueExtractor.extract(actualResult, ((ExpectedResult) expectedResult).getProperty());
        } else {
            return actualResult;
        }
    }

    private void constructErrorMessage(ExpectedResult expectedResult, Object actualValue, StringBuilder errorBuilder) {
        errorBuilder.append(expectedResult.getFullPropertyPath())
                .append(" is not valid. ")
                .append(expectedResult.getValidationTypeOfParent() != ValidationType.NONE_MATCH ? "Expected: " : "Did not expect: ")
                .append(expectedResult.<Object>getExpectedResult())
                .append(" Actual: ")
                .append(actualValue)
                .append('\n');
    }

    public static class Builder {

        private ValueComparator valueComparator;
        private TestStepValueExtractor valueExtractor;

        public Builder withValueComparator(ValueComparator valueComparator) {
            this.valueComparator = valueComparator;
            return this;
        }

        public Builder withValueExtractor(TestStepValueExtractor valueExtractor) {
            this.valueExtractor = valueExtractor;
            return this;
        }

        public TestStepResultValidator build() {
            return new DefaultTestStepResultValidator(valueComparator, valueExtractor);
        }
    }
}

File: skellig-test-step-processing/src/main/kotlin/org/skellig/teststep/processing/validation/DefaultTestStepResultValidator.kt
package org.skellig.teststep.processing.validation

import org.skellig.teststep.processing.exception.ValidationException
import org.skellig.teststep.processing.model.ExpectedResult
import org.skellig.teststep.processing.model.MatchingType
import org.skellig.teststep.processing.value.chunk.RawValueProcessingVisitor
import org.skellig.teststep.processing.value.chunk.RawValueChunk

class DefaultTestStepResultValidator(
    private val rawValueProcessingVisitor: RawValueProcessingVisitor
) : TestStepResultValidator {

    override fun validate(expectedResult: ExpectedResult, actualResult: Any?) {
        val errorBuilder = StringBuilder()
        if (!validate(expectedResult, actualResult, errorBuilder)) {
            throw ValidationException("Validation failed!\n$errorBuilder")
        }
    }

    private fun validate(expectedResult: ExpectedResult, actualResult: Any?, errorBuilder: StringBuilder): Boolean {
        return when (expectedResult.matchingType) {
            MatchingType.ANY_MATCH -> {
                expectedResult.get<List<ExpectedResult>>()!!
                    .any { item: ExpectedResult -> validateFurther(item, actualResult, errorBuilder) }
            }
            MatchingType.ANY_NONE_MATCH -> {
                expectedResult.get<List<ExpectedResult>>()!!
                    .any { item: ExpectedResult -> !validateFurther(item, actualResult, errorBuilder) }
            }
            MatchingType.NONE_MATCH -> {
                expectedResult.get<List<ExpectedResult>>()!!
                    .none { item: ExpectedResult -> validateFurther(item, actualResult, errorBuilder) }
            }
            MatchingType.ALL_MATCH -> {
                expectedResult.get<List<ExpectedResult>>()!!
                    .all { item: ExpectedResult -> validateFurther(item, actualResult, errorBuilder) }
            }
            else -> {
                val isValid = rawValueProcessingVisitor.process(expectedResult.expectedResult as RawValueChunk?, actualResult)

                if (expectedResult.getMatchingTypeOfParent() === MatchingType.NONE_MATCH && isValid ||
                    expectedResult.getMatchingTypeOfParent() !== MatchingType.NONE_MATCH && !isValid
                ) {
                    constructErrorMessage(expectedResult, actualResult, errorBuilder)
                }
                isValid
            }
        }
    }

    private fun validateFurther(expectedResult: ExpectedResult, actualResult: Any?, errorBuilder: StringBuilder): Boolean {
        val newActualValue = extractActualValueFromExpectedResult(actualResult, expectedResult)
        return if (expectedResult.isGroup()) {
            validate(expectedResult, newActualValue, errorBuilder)
        } else {
            if (expectedResult.property == null && newActualValue is Collection<*>) {
                if (expectedResult.matchingType === MatchingType.NONE_MATCH) {
                    newActualValue.all { validate(expectedResult, it, errorBuilder) }
                } else {
                    newActualValue.any { validate(expectedResult, it, errorBuilder) }
                }
            } else if (newActualValue?.javaClass?.isArray == true && newActualValue !is ByteArray) {
                validateFurther(expectedResult, (newActualValue as Array<*>).toList(), errorBuilder)
            } else {
                validate(expectedResult, newActualValue, errorBuilder)
            }
        }
    }

    private fun extractActualValueFromExpectedResult(actualResult: Any?, expectedResult: Any): Any? {
        return if (ExpectedResult::class.java == expectedResult.javaClass && (expectedResult as ExpectedResult).property != null) {
            rawValueProcessingVisitor.process(actualResult, expectedResult.property)
        } else {
            actualResult
        }
    }

    private fun constructErrorMessage(expectedResult: ExpectedResult, actualValue: Any?, errorBuilder: StringBuilder) {
        errorBuilder.append(expectedResult.getFullPropertyPath())
            .append(" is not valid. ")
            .append(if (expectedResult.getMatchingTypeOfParent() !== MatchingType.NONE_MATCH) "Expected: " else "Did not expect: ")
            .append(expectedResult.expectedResult)
            .append(" Actual: ")
            .append(actualValue)
            .append('\n')
    }

    class Builder {
        private var rawValueProcessingVisitor: RawValueProcessingVisitor? = null

        fun withValueProcessingVisitor(rawValueProcessingVisitor: RawValueProcessingVisitor?) = apply {
            this.rawValueProcessingVisitor = rawValueProcessingVisitor
        }

        fun build(): TestStepResultValidator {
            return DefaultTestStepResultValidator(rawValueProcessingVisitor!!)
        }
    }
}